### ISSUE TITLE
Upgarding from 1.x to 2.x

### ISSUE THREAD

**Issue by ahwayakchih** [symphonists/subsectionmanager#127]
I'm currently fighting with upgrade. It's not as painless as it could be so i'll be posting about problems here. Hopefully i can either find a fix or @nilshoerrmann can explain me where i did something wrong :).
1. As mentioned at https://github.com/nilshoerrmann/subsectionmanager/issues/117#issuecomment-1326278 there is no quick way to move section field settings to data source settings. It would be great to have some kind of a one-click-to-import way that would select output fields based on settings from 1.x. Selecting correct fields from over 1000 (one thousand) items is not much fun, especially when it is just first data source to update from total of about 20 :).
2. As mentioned at https://github.com/nilshoerrmann/subsectionmanager/issues/117#issuecomment-1326278 i worry about Reflection field being able to access data of subsection manager field entries. I guess that Reflection field should allow user to select data source from which it will get entry data?
3. When selected data source output fields like "Objects: Photos: Photo" (upload field in "Photos" section, that is pointed to by "Objects" section's SSM) XML did not output photo. It stopped on empty item tag with id:

```
    <objects id="166" items="1">
        <item id="15">
             <photos id="61" items="4">
                 <item id="5" />
             </photos>
        </item>
    </objects>
```

Is it supposed to be like that, or should it output photos/item/photo too?

_update_: i actually counted field items in data source. It is not "close to one hundred", it is "1011" :)!

**Comment by nilshoerrmann** [comment]
1. SSM 2 is a development version which is the reason why no upgrade script exist yet to solve this.
2. I'm not sure about this. Have you tried this out or is it just a guess that it doesn't work?
3. Are we talking about nested Subsection Managers? It should work (at least [this](https://github.com/nilshoerrmann/subsectionmanager/commit/b39f3597def86b60dcabe5947e22cba2d598a80c) should have fixed it).

**Comment by ahwayakchih** [comment]
1. I know. Sorry if it sounded harsh, i just wanted to know Your opinion on the idea of a JavaScript solution.
2. I cannot try it yet, as nothing works right now. It is a guess based on knowledge that SSM needs Data Source to load data from subsection fileds. Reflection field does not use any Data Sources (it simply calls `appendFormattedElement` on each of `$entry` fields).
3. Yes. "root" section has SSM field pointing to "child" section that has SSM field pointing to... etc... (that is why i discovered recursion problems, after a few child->child steps it can get back to the "root" section :).

**Comment by nilshoerrmann** [comment]
Is it possible that this is a quite complex site you are developing?

> It is a guess based on knowledge that SSM needs Data Source to load data from subsection fileds.

Why does SSM need a data source to load it's subsection items?!

**Comment by ahwayakchih** [comment]
> Is it possible that this is a quite complex site you are developing?

Well... it is not a simple site for sure :).
That's good for SSM - the more crazy environments it will be tested in, the more bugs will be fixed.

> Why does SSM need a data source to load it's subsection items?!

It does not? Since i have to set fields in Data Source and "selected fields" are not on section edit page anymore... i assumed it requires some Data Source. Looks like i was wrong ;).

> It should work (at least this should have fixed it).

Hmm... that is strange. 2.x branch does not seem to have that fix included:
https://github.com/nilshoerrmann/subsectionmanager/blob/2.x/extension.driver.php#L216

**Comment by ahwayakchih** [comment]
It looks like https://github.com/nilshoerrmann/subsectionmanager/commit/b39f3597def86b60dcabe5947e22cba2d598a80c was reversed by https://github.com/nilshoerrmann/subsectionmanager/commit/9acb67ea71ec7ff3cb1b3396ee6d452a5661142a

**Comment by nilshoerrmann** [comment]
I just upgraded an install I'm working on and I can reproduce your third issue (the missing entry output). I'll investigate ...

**Comment by nilshoerrmann** [comment]
Could you please have a look at this commit: https://github.com/nilshoerrmann/subsectionmanager/commit/e6ddcda09f652595286b644e0d9fa9ee6f1eb993

**Comment by ahwayakchih** [comment]
It did not help :(

```
                    <photos id="61" items="4">
                        <item id="5" />
                        <item id="6" />
                        <item id="7" />
                        <item id="60" />
                        <item id="61" />
                    </photos>
```

I have selected `ssm : photos : photo` (file upload field), `ssm : photos : name : unformatted` (text box field), and `ssm : photos : description : formatted` (another text box field) in Data Source.

**Comment by nilshoerrmann** [comment]
In this case I'm stuck - it's working perfectly here.

**Comment by nilshoerrmann** [comment]
Is it possible that this issue is related to your latest comment [here](https://github.com/nilshoerrmann/subsectionmanager/pull/128)? 

It looks a bit strange that you have a count of 4 items but 5 attached nodes where the last has the same id as the parent `photos` entry.

**Comment by ahwayakchih** [comment]
I thought so, but other fields do not show items either, even though none of their names is the same as other field.

It indeed does look strange. I just checked other fields, and every one of them lists one last item with ID the same as field ID.
`photos` is not an entry, it is SSM field. It looks like this:

```
    <reviews>
        <section id="1" handle="reviews">Reviews</section>
        <entry id="1">
            <title handle="blah-test">Blah test</title>
            <date time="11:53" weekday="6">2011-04-30</date>
            <photos id="18" items="3">
                <item id="17">
                    <photo size="81 KB" path="/uploads" type="image/jpeg">
                        <filename>test.jpg</filename>
                        <meta creation="2011-03-27T23:41:17+02:00" width="577" height="479" />
                    </photo>
                    <description mode="formatted" word-count="0" />
                    <name mode="unformatted" word-count="0"><![CDATA[]]></name>
                </item>
                /* rest of items here */
            </photos>
            <languages>
                <item handle="pl">polski</item>
            </languages>
            <objects id="166" items="1">
                <item id="15">
                    <name handle="whatever">whatever</name>
                    <type>
                        <item handle="block">Block</item>
                    </type>
                    <series mode="unformatted" handle="lalala" word-count="0"><![CDATA[LALALA]]></series>
                    <model mode="unformatted" handle="" word-count="0"><![CDATA[]]></model>
                    <specifications id="432" items="2">
                        <item id="31" />
                        <item id="30" />
                        <item id="432" />
                    </specifications>
                    <notes id="714" items="2">
                        <item id="713" />
                        <item id="714" />
                    </notes>
                    <awards items="0" />
                    <photos id="61" items="4">
                        <item id="5" />
                        <item id="6" />
                        <item id="7" />
                        <item id="60" />
                        <item id="61" />
                    </photos>
                </item>
                /* ... */
            </objects id="166" items="1">
        </entry>
    </reviews>
```

Unless it is a part of the problem because Objects have "Parts" SSM field pointing to Objects? But why it would affect Photos? Especially if i did not select any of "Parts" fields in Data Source.

**Comment by ahwayakchih** [comment]
Just checked that `reviews/entry/photos/@id` of last item is the same as `reviews/entry/photos/@id`. It looks like it outputs ID of last item, not ID of field.

**Comment by ahwayakchih** [comment]
Count items may be less than actual items, because of another bug: https://github.com/nilshoerrmann/subsectionmanager/issues/120

_update_: yep, Photo with ID=5 is not there (entry was deleted).
_update_: Not exsiting photos fixed in https://github.com/nilshoerrmann/subsectionmanager/pull/129

**Comment by ahwayakchih** [comment]
I got part of nested SSM fields problem fixed. Still working on fixing rest of it :).

**Comment by ahwayakchih** [comment]
1. Raw and experimental implementation at https://github.com/ahwayakchih/subsectionmanager/commit/8a2b269a6137f9a451c64fc24a808d1e6bb145fa
   I had to import settings fast, so did not care much about interface or clean code. I think code should be in separate function instead (maybe ajax page loaded on demand).
2. As i expected, Reflection field is not able to use subsection data anymore. SubsectionField 2.x requires data source to know which fields to load. So Reflection field has to be updated and either "fake" data source, or... probably best will be to make it possible to load whole entry XML from URL. I do something similar in Autocompleter extension. It would actually fix another problem in Reflection field - to be able to access old value stored in entry. Developers will be able to decide what data is passed to Reflection field the same way they decide what to output on frontpage :).
   I think the same problem applies to entry_url extension (and any other that depends on `appendFormattedElement`). Maybe SubsectionManager should output all fields data by default? It could check if Data Source was called or not.
   _update_: fixed in https://github.com/nilshoerrmann/subsectionmanager/pull/131
3. Fixed in https://github.com/nilshoerrmann/subsectionmanager/pull/130
